### PR TITLE
[Foundation] Add support for Xcode 15 beta 7.

### PR DESCRIPTION
**Pull request by mandel-macaque**

### CODE REVIEW

File: src/Foundation/Enum.cs [hunk 1]
```diff
@@ -1380,6 +1385,9 @@ public enum NSFileProtectionType {
 		CompleteUnlessOpen,
 		[Field ("NSFileProtectionCompleteUntilFirstUserAuthentication")]
 		CompleteUntilFirstUserAuthentication,
+		[Watch (10, 0), TV (17, 0), NoMac, iOS (17, 0), MacCatalyst (17,0)]
+		[Field ("NSFileProtectionCompleteWhenUserInactive")]
+		CompleteWhenUserInactive,
```
Comment (mandel-macaque) [inline src/Foundation/Enum.cs:1390]: This, is a normal situation, should not be allowed. YET! this enum was never released and was added in https://github.com/xamarin/xamarin-macios/pull/18918 meaning no developer ever has used this enum.

File: src/foundation.cs [hunk 2]
```diff
@@ -875,8 +875,76 @@ public enum NSAttributedStringNameKey {
...
 		[Field ("NSReplacementIndexAttributeName")]
 		ReplacementIndex,
+
+		[Watch (10, 0), TV (17, 0), Mac (14, 0), iOS (17, 0), MacCatalyst (17, 0)]
+		[Field ("NSInflectionAgreementArgumentAttributeName")]
+		InflectionAgreementArgument,
+
+		[Watch (10, 0), TV (17, 0), Mac (14, 0), iOS (17, 0), MacCatalyst (17, 0)]
+		[Field ("NSInflectionAgreementConceptAttributeName")]
+		InflectionConcept,
```
Comment (rolfbjarne) [inline src/foundation.cs:885]: This matches the naming for the previous value

```suggestion
		InflectionAgreementConcept,
```

File: src/foundation.cs [hunk 3]
```diff
@@ -875,8 +875,76 @@ public enum NSAttributedStringNameKey {
...
+
+		[Watch (10, 0), TV (17, 0), Mac (14, 0), iOS (17, 0), MacCatalyst (17, 0)]
+		[Field ("NSInflectionReferentConceptAttributeName")]
+		InflectionReferentConceptAttributeName,
```
Comment (rolfbjarne) [inline src/foundation.cs:889]: Same
```suggestion
		InflectionReferentConcept,
```

File: src/foundation.cs [hunk 4]
```diff
@@ -875,8 +875,76 @@ public enum NSAttributedStringNameKey {
...
+	[Watch (10, 0), TV (17, 0), Mac (14, 0), iOS (17, 0), MacCatalyst (17, 0)]
+	[Native]
+	public enum NSGrammaticalPerson : long {
+		NotSet = 0,
+		First,
+		Second,
+		Third,
 	}
 
+
```
Comment (rolfbjarne) [inline src/foundation.cs:947]:
```suggestion
```

File: src/foundation.cs [hunk 5]
```diff
@@ -6183,6 +6263,12 @@ partial interface NSUrl : NSSecureCoding, NSCopying
 		[Static]
 		NSUrl _FromStringRelative (string url, NSUrl relative);
 
+		[Watch (10, 0), TV (17, 0), Mac (14, 0), iOS (17, 0), MacCatalyst (17, 0)]
+		[Static]
+		[Export ("URLWithString:encodingInvalidCharacters:")]
+		[return: NullAllowed]
+		NSUrl FromStringEncoded (string url, bool encodingInvalidCharacters);
```
Comment (rolfbjarne) [inline src/foundation.cs:6270]:
```suggestion
		NSUrl FromString (string url, bool encodingInvalidCharacters);
```

File: src/foundation.cs [hunk 6]
```diff
@@ -6897,6 +6983,14 @@ partial interface NSUrl : NSSecureCoding, NSCopying
 		[Field ("NSURLFileProtectionCompleteUntilFirstUserAuthentication")]
 		NSString FileProtectionCompleteUntilFirstUserAuthentication { get; }
 
+		[Watch (10, 0), TV (17, 0), NoMac, iOS (17, 0), MacCatalyst (17, 0)]
+		[Field ("NSURLFileProtectionCompleteWhenUserInactive")]
+		NSString FileProtectionCompleteWhenUserInactive { get; }
+
+		[Watch (10, 0), TV (17, 0), Mac (14, 0), iOS (17, 0), MacCatalyst (17, 0)]
+		[Field ("NSURLDirectoryEntryCountKey")]
+		NSString NSURLDirectoryEntryCountKey { get; }
```
Comment (rolfbjarne) [inline src/foundation.cs:6992]:
```suggestion
		NSString DirectoryEntryCountKey { get; }
```

File: src/foundation.cs [hunk 7]
```diff
@@ -7096,6 +7190,12 @@ partial interface NSUrlComponents : NSCopying {
 		[Static, Export ("componentsWithString:")]
 		NSUrlComponents FromString (string urlString);
 
+		[Watch (10, 0), TV (17, 0), Mac (14, 0), iOS (17, 0), MacCatalyst (17, 0)]
+		[Static]
+		[Export ("componentsWithString:encodingInvalidCharacters:")]
+		[return: NullAllowed]
+		NSUrlComponents FromStringEncoded (string url, bool encodingInvalidCharacters);
```
Comment (rolfbjarne) [inline src/foundation.cs:7197]:
```suggestion
		NSUrlComponents FromString (string url, bool encodingInvalidCharacters);
```

File: src/foundation.cs [hunk 8]
```diff
@@ -7743,6 +7843,15 @@ partial interface NSUrlSession {
 		[MacCatalyst (13, 1)]
 		[Export ("webSocketTaskWithRequest:")]
 		NSUrlSessionWebSocketTask CreateWebSocketTask (NSUrlRequest request);
+
+		[Watch (10, 0), TV (17, 0), Mac (14, 0), iOS (17, 0), MacCatalyst (17, 0)]
+		[Export ("uploadTaskWithResumeData:")]
+		NSUrlSessionUploadTask UploadTask (NSData resumeData);
```
Comment (rolfbjarne) [inline src/foundation.cs:7849]:
```suggestion
		NSUrlSessionUploadTask CreateUploadTask (NSData resumeData);
```

File: src/foundation.cs [hunk 9]
```diff
@@ -7743,6 +7843,15 @@ partial interface NSUrlSession {
 		[MacCatalyst (13, 1)]
 		[Export ("webSocketTaskWithRequest:")]
 		NSUrlSessionWebSocketTask CreateWebSocketTask (NSUrlRequest request);
+
+		[Watch (10, 0), TV (17, 0), Mac (14, 0), iOS (17, 0), MacCatalyst (17, 0)]
+		[Export ("uploadTaskWithResumeData:")]
```
Comment (rolfbjarne) [inline src/foundation.cs:7848]: The return type probably needs the ForcedType attribute, like the other methods in this type:

```suggestion
		[Export ("uploadTaskWithResumeData:")]
		[return: ForcedType]
```

File: src/foundation.cs [hunk 10]
```diff
@@ -7743,6 +7843,15 @@ partial interface NSUrlSession {
...
+		NSUrlSessionUploadTask UploadTask (NSData resumeData);
+
+		[Async (ResultTypeName = "NSUrlSessionUploadTaskResumeRequest")]
+		[Watch (10, 0), TV (17, 0), Mac (14, 0), iOS (17, 0), MacCatalyst (17, 0)]
+		[Export ("uploadTaskWithResumeData:completionHandler:")]
```
Comment (rolfbjarne) [inline src/foundation.cs:7853]:
```suggestion
		[Export ("uploadTaskWithResumeData:completionHandler:")]
		[return: ForcedType]
```

File: src/foundation.cs [hunk 11]
```diff
@@ -7743,6 +7843,15 @@ partial interface NSUrlSession {
...
+		NSUrlSessionUploadTask UploadTask (NSData resumeData, Action<NSData, NSUrlResponse, NSError> completionHandler);
```
Comment (rolfbjarne) [inline src/foundation.cs:7854]:
```suggestion
		NSUrlSessionUploadTask CreateUploadTask (NSData resumeData, Action<NSData, NSUrlResponse, NSError> completionHandler);
```

File: src/foundation.cs [hunk 12]
```diff
@@ -7930,6 +8039,15 @@ partial interface NSUrlSessionUploadTask {
 		[Deprecated (PlatformName.MacCatalyst, 13, 1, message: "Use 'NSURLSession.CreateUploadTask' instead.")]
 		[Export ("init")]
 		NativeHandle Constructor ();
+
+		[Watch (10, 0), TV (17, 0), Mac (14, 0), iOS (17, 0), MacCatalyst (17, 0)]
+		[Field ("NSURLSessionUploadTaskResumeData")]
+		NSString TaskResumeDataKey { get; }
```
Comment (rolfbjarne) [inline src/foundation.cs:8045]:
```suggestion
		NSString ResumeDataKey { get; }
```

File: src/foundation.cs [hunk 13]
```diff
@@ -8196,6 +8314,14 @@ partial interface NSUrlSessionTaskDelegate {
...
 		[Export ("URLSession:didCreateTask:")]
 		void DidCreateTask (NSUrlSession session, NSUrlSessionTask task);
+
+		[Watch (10, 0), TV (17, 0), Mac (14, 0), iOS (17, 0), MacCatalyst (17, 0)]
+		[Export ("URLSession:task:didReceiveInformationalResponse:")]
+		void DidReceiveInformationalResponse (NSUrlSession session, NSUrlSessionTask task, NSHttpUrlResponse response);
+
+		[Watch (10, 0), TV (17, 0), Mac (14, 0), iOS (17, 0), MacCatalyst (17, 0)]
+		[Export ("URLSession:task:needNewBodyStreamFromOffset:completionHandler:")]
+		void NeedNewBodyStreamFromOffset (NSUrlSession session, NSUrlSessionTask task, long offset, Action<NSInputStream> completionHandler);
```
Comment (rolfbjarne) [inline src/foundation.cs:8324]:
```suggestion
		void NeedNewBodyStream (NSUrlSession session, NSUrlSessionTask task, long offset, Action<NSInputStream> completionHandler);
```

File: src/foundation.cs [hunk 14]
```diff
@@ -17874,10 +18000,18 @@ interface NSMorphology : NSCopying, NSSecureCoding {
 		[Export ("number", ArgumentSemantic.Assign)]
 		NSGrammaticalNumber Number { get; set; }
 
+		[Obsoleted (PlatformName.MacOSX, 14, 0, message: "Use 'NSTermOfAddress' instead.")]
+		[Obsoleted (PlatformName.iOS, 17, 0, message: "Use 'NSTermOfAddress' instead.")]
```
Comment (rolfbjarne) [inline src/foundation.cs:18004]:
```suggestion
		[Obsoleted (PlatformName.iOS, 17, 0, message: "Use 'NSTermOfAddress' instead.")]
		[Obsoleted (PlatformName.MacCatalyst, 17, 0, message: "Use 'NSTermOfAddress' instead.")]
```

File: src/foundation.cs [hunk 15]
```diff
@@ -17874,10 +18000,18 @@ interface NSMorphology : NSCopying, NSSecureCoding {
...
+		[Obsoleted (PlatformName.TvOS, 17, 0, message: "Use 'NSTermOfAddress' instead.")]
+		[Obsoleted (PlatformName.WatchOS, 10, 0, message: "Use 'NSTermOfAddress' instead.")]
 		[Export ("customPronounForLanguage:")]
 		[return: NullAllowed]
 		NSMorphologyCustomPronoun GetCustomPronoun (string language);
 
+		[Obsoleted (PlatformName.MacOSX, 14, 0, message: "Use 'NSTermOfAddress' instead.")]
+		[Obsoleted (PlatformName.iOS, 17, 0, message: "Use 'NSTermOfAddress' instead.")]
```
Comment (rolfbjarne) [inline src/foundation.cs:18012]:
```suggestion
		[Obsoleted (PlatformName.iOS, 17, 0, message: "Use 'NSTermOfAddress' instead.")]
		[Obsoleted (PlatformName.MacCatalyst, 17, 0, message: "Use 'NSTermOfAddress' instead.")]
```

File: src/foundation.cs [hunk 16]
```diff
@@ -17887,8 +18021,32 @@ interface NSMorphology : NSCopying, NSSecureCoding {
...
+		[Export ("pronounType", ArgumentSemantic.Assign)]
+		NSGrammaticalPronounType PronounType { get; set; }
+
+		[Watch (10, 0), TV (17, 0), Mac (14, 0), iOS (17, 0), MacCatalyst (17, 0)]
+		[Export ("definiteness", ArgumentSemantic.Assign)]
+		NSGrammaticalDefiniteness Definiteness { get; set; }
 	}
 
+	[Obsoleted (PlatformName.MacOSX, 14, 0, message: "Use 'NSTermOfAddress' instead.")]
+	[Obsoleted (PlatformName.iOS, 17, 0, message: "Use 'NSTermOfAddress' instead.")]
```
Comment (rolfbjarne) [inline src/foundation.cs:18047]:
```suggestion
	[Obsoleted (PlatformName.iOS, 17, 0, message: "Use 'NSTermOfAddress' instead.")]
	[Obsoleted (PlatformName.MacCatalyst, 17, 0, message: "Use 'NSTermOfAddress' instead.")]
```

**Review by rolfbjarne — CHANGES_REQUESTED** [review]

File: src/foundation.cs (hunk 5 above)
Comment (mandel-macaque) [reply src/foundation.cs:6270]: The is a reason for this, there is a method already with the same parameters as that same name.

File: src/foundation.cs (hunk 7 above)
Comment (mandel-macaque) [reply src/foundation.cs:7197]: Same

File: src/foundation.cs (hunk 5 above)
Comment (rolfbjarne) [reply src/foundation.cs:6269]: There's a constructor with the same parameters, but not a method as far as I can tell?

**Review by dalexsoto — APPROVED** [review]
After @rolfbjarne 👍

File: src/foundation.cs [hunk 17]
```diff
@@ -8196,6 +8315,14 @@ partial interface NSUrlSessionTaskDelegate {
...
 		[Export ("URLSession:didCreateTask:")]
 		void DidCreateTask (NSUrlSession session, NSUrlSessionTask task);
+
+		[Watch (10, 0), TV (17, 0), Mac (14, 0), iOS (17, 0), MacCatalyst (17, 0)]
+		[Export ("URLSession:task:didReceiveInformationalResponse:")]
+		void DidReceiveInformationalResponse (NSUrlSession session, NSUrlSessionTask task, NSHttpUrlResponse response);
+
+		[Watch (10, 0), TV (17, 0), Mac (14, 0), iOS (17, 0), MacCatalyst (17, 0)]
+		[Export ("URLSession:task:needNewBodyStreamFromOffset:completionHandler:")]
+		void NeedNewBodyStream (NSUrlSession session, NSUrlSessionTask task, long offset, Action<NSInputStream> completionHandler);
```
Comment (tj-devel709) [inline src/foundation.cs:8325]: Maybe Async?

**Review by tj-devel709 — APPROVED** [review]

**Comment by mandel-macaque** [comment]
@rolfbjarne I implemented the requested changes and tests pass.

File: src/Foundation/Enum.cs [hunk 18]
```diff
@@ -1380,6 +1385,9 @@ public enum NSFileProtectionType {
 		CompleteUnlessOpen,
 		[Field ("NSFileProtectionCompleteUntilFirstUserAuthentication")]
 		CompleteUntilFirstUserAuthentication,
+		[Watch (10, 0), TV (17, 0), NoMac, iOS (17, 0), MacCatalyst (17, 0)]
+		[Field ("NSFileProtectionCompleteWhenUserInactive")]
+		CompleteWhenUserInactive,
 		[Field ("NSFileProtectionNone")]
 		None,
```
Comment (dalexsoto) [inline src/Foundation/Enum.cs:1392]: There is a breaking change

image


```suggestion
		[Field ("NSFileProtectionNone")]
		None,
		[Watch (10, 0), TV (17, 0), NoMac, iOS (17, 0), MacCatalyst (17, 0)]
		[Field ("NSFileProtectionCompleteWhenUserInactive")]
		CompleteWhenUserInactive,
```

**Review by dalexsoto — CHANGES_REQUESTED** [review]

File: src/foundation.cs (hunk 17 above)
Comment (rolfbjarne) [reply src/foundation.cs:8325]: @tj-devel709 This is a protocol, we can't use Async in protocols.

**Review by rolfbjarne — CHANGES_REQUESTED** [review]
The PR must be retargeted to `main`, otherwise this looks good now (except the breaking change Alex mentioned).

**Review by tj-devel709 — APPROVED** [review]

**Review by dustin-wojciechowski — APPROVED** [review]

**Review by dalexsoto — APPROVED** [review]

**Review by rolfbjarne — APPROVED** [review]

**Pull request merged**